### PR TITLE
Various HVACTemplate fixes including new fan coil control and coil options

### LINKED ISSUE #5260: HVACTemplate:Zone:FanCoil add support for new electric heating coil and multispeed fan control
Helpdesk ticket 10837
v8.4.0 added a new electric heating coil to ZoneHVAC:FourPipeFanCoil and a new capacity control type MultiSpeedFan.  Add support for these in HVACTemplate:Zone:FanCoil.

Test files [5260-HVACTemplateFanCoil.zip](https://github.com/NREL/EnergyPlus/files/2641127/5260-HVACTemplateFanCoil.zip)

### LINKED ISSUE #5617: AirTerminal:SingleDuct:ParallelPIU:Reheat Secondary Supply Air Maximum Flow Rate sizes to cooling flow - should it be heating flow?
## Issue overview

User noted for HVACTemplate:Zone:VAV:FanPowered if a hard value is specified for the Secondary Supply Air Maximum Flow Rate, this value is passed as the Design Heating Flow Rate in Sizing:Zone, but the AirTerminal:SingleDuct:ParallelPIU:Reheat object sizes the secondary flow rate to match the cooling flow, not this heating flow.  This is for a parallel PIU.  So, should this value simply be placed directly into the terminal unit, or does the autosizing need to be changed to work off the heating design flow?
### Workaround

Use the expidf output file, copy to <newname>.idf, and enter the hard value flow rate(s) there.
### Details

Some additional details for this issue (if relevant):
- Win64
- EnergyPlus v8.5.0
- Helpdesk ticket 11278
### Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Defect file added (EnergyPlusDevSupport\DefectFiles\5617)
- [ ] Ticket added to Pivotal for defect (development team task)
- [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #7047: HVACTemplate:System:DedicatedOutdoorAir dx cooling coil significantly undersized.
Issue overview
--------------
`HVACTemplate:System:DedicatedOutdoorAir` generates a `CoilSystem:Cooling:DX` object but fails to add the "Use Outdoor Air DX Coil" field, so for a climate in Indonesia, the coil is significantly undersized and unable to meet the 12.8C cooling setpoint.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus v9.0.1
 - Unmethours [35070](https://unmethours.com/question/35070/how-e-calculate-the-ventilation-loads-and-what-is-the-output/?comment=35380#post-id-35380)

### Workaround
Take the expidf output, rename to <newfilename>.idf and add the missing field to this object (once for each HVACTemplate DOAS system), remembering to change the semi-colon to a comma on the "Run on Latent Load" line:
```
CoilSystem:Cooling:DX,
  DOAS DX Cooling System,                                  !- Name
  ,                                                        !- Availability Schedule Name
  DOAS Supply Fan Outlet,                                  !- DX Cooling Coil System Inlet Node Name
  DOAS Cooling Coil Outlet,                                !- DX Cooling Coil System Outlet Node Name
  DOAS Cooling Coil Outlet,                                !- DX Cooling Coil System Sensor Node Name
  Coil:Cooling:DX:TwoStageWithHumidityControlMode,         !- Cooling Coil Object Type
  DOAS Cooling Coil,                                       !- Cooling Coil Name
  None,                                                    !- Dehumidification Control Type
  Yes,                                                     !- Run on Sensible Load
  No,                                                      !- Run on Latent Load
  Yes;                                                     !- Use Outdoor Air DX Coil
```

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #7061: HVACTemplate:System:ConstantVolume Minimum Outdoor Air Schedule Name is ignored
Issue overview
--------------
HVACTemplate:System:ConstantVolume Minimum Outdoor Air Schedule Name is ignored.

### Workaround
The only way around this is to edit the expidf file, or to use a different system type, such as `HVACTemplate:System:UnitarySystem` or `HVACTemplate:System:VAV` with the terminal unit min fractions set to 1.0.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus v9.0.1
 - Unmethours [35602](https://unmethours.com/question/35602/hvactemplatesystemconstantvolume-minimum-outdoor-air-schedule-doesnt-work/).

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
- [x] Fixes #7061, HVACTemplate:System:ConstantVolume "Minimum Outdoor Air Schedule Name is ignored".

- [x] Fixes #7047, DOAS DX coil undersized.

- [x] Fixes #5260, in `HVACTemplate:Zone:FanCoil` add new fan coil control options and electric heating coil. 
Also changed this to use `Fan:SystemModel` for all control types.
Also added cooling coil type "HeatExchangerAssistedChilledWater".

### Diffs Review
**HVACTemplate-5ZoneFanCoil**, **HAMT_DailyProfileReport** and **HAMT_HourlyProfileReport**: eio diffs because of Fan:SystemModel instead of Fan:OnOff

**HVACTemplate-5ZoneFanCoil-DOAS**: eio diffs because of fan type, but unexpected eso big diffs, punting to #7086 

**HVACTemplate-5ZonePTAC-DOAS**: eio diffs and eso big diffs expected because the DOAS DX cooling coil is now sized as a 100% OA coil (allows lower cfm/ton).

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Functional code review (it has to work!)
 - [x] If defect, results of running current develop vs this branch should exhibit the fix
 - [x] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check
 - [ ] Unit Test(s)
 - C++ checks:
   - [ ] Argument types
   - [ ] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [ ] Verify naming conventions and styles, memos and notes and defaults
   - [ ] Open windows IDF Editor with modified IDD to check for errors
   - [ ] If transition, add to input rules file for interfaces
   - [ ] If transition, add transition source
   - [ ] If transition, update idfs
 - [ ] If new idf included, locally check the err file and other outputs
 - [ ] Required documentation updates?
 - [ ] ExpandObjects changes?
 - [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

**Comment by Myoldmopar** [comment]
@mjwitte there's a duplicate Marvin result there because I was alerted that the regression files weren't being posted.  I fixed that and re-ran CI on this branch with Marvin and they now look great again.  Carry on...

**Comment by mjwitte** [comment]
Thanks! Happy New Year! @nealkruis @jasondegraw Diff results are fixed.

File: doc/input-output-reference/src/hvac-template-objects/group-hvac-templates.tex [hunk 1]
```diff
@@ -667,17 +667,19 @@ \subsubsection{Inputs}\label{inputs-3-018}
...
 
 \begin{itemize}
 \item
-  ChilledWater
+  \emph{ChilledWater} uses the Coil:Cooling:Water object with the DetailedAnalysis option. This is the preferred coil model for most applications.
 \item
-  ChilledWaterDetailedFlatModel
+   \emph{ChilledWaterDetailedFlatModel} uses the Coil:Cooling:Water:DetailedGeometry object. This option is available for special applications.
+\item
+   \emph{HeatExchangerAssistedChilledWater} uses CoilSystem:Cooling:Water:HeatExchangerAssisted with Coil:Cooling:Water for enhanced dehumidification.
```
Comment (Myoldmopar) [inline doc/input-output-reference/src/hvac-template-objects/group-hvac-templates.tex:678]: Random note here: LaTeX offers a `description` environment which is a nice fit for this sort of keyword-description list.  No worries here, just feel free to have it as a possibility for later.

File: doc/input-output-reference/src/hvac-template-objects/group-hvac-templates.tex [hunk 2]
```diff
@@ -736,7 +730,23 @@ \subsubsection{Inputs}\label{inputs-3-018}
...
+  \emph{CyclingFan}: The fan speed is chosen so that the unit capacity is greater than or equal to the heating / cooling load and the fan is cycled to match unit output with the load.
+\item
+  \emph{VariableFanVariableFlow}: Both air and water flow rates are varied to match the load.
+\item
+  \emph{VariableFanConstantFlow}: The water flow rate is at full flow and the fan speed varies to meet the load.
+\item
+  \emph{MultiSpeedFan}: The water flow rate is at full flow when there is load or fully closed when there is no load and the supply air flow rate is varied by varying the fan speed in order to match the load.
+\item
+  \emph{ASHRAE90VariableFan}: The fan air flow rate is reduced according to the low speed supply air flow ratio when the zone sensible load is less than the zone sensible load multiplied by the low speed supply air flow ratio. 
+\end{itemize}
```
Comment (Myoldmopar) [inline doc/input-output-reference/src/hvac-template-objects/group-hvac-templates.tex:747]: Much nicer.

File: src/ExpandObjects/epfilter.f90 [hunk 3]
```diff
@@ -23676,24 +23686,26 @@ SUBROUTINE CreateFanCoil
...
   CALL AddToObjStr('Outdoor Air Mixer Object Type', 'OutdoorAir:Mixer')
   CALL AddToObjFld('Outdoor Air Mixer Name', base + fczNameOff,' OA Mixing Box')
-  IF (capacityControlKind == capctrlConstFanVarFlow .OR. capacityControlKind==capctrlASHRAE90VarFan) THEN
-    CALL AddToObjStr('Supply Air Fan Object Type', 'Fan:ConstantVolume')
-  ELSEIF (capacityControlKind == capctrlCyclFan) THEN
-    CALL AddToObjStr('Supply Air Fan Object Type', 'Fan:OnOff')
-  ELSE
-    CALL AddToObjStr('Supply Air Fan Object Type', 'Fan:VariableVolume')
-  ENDIF
+  CALL AddToObjStr('Supply Air Fan Object Type', 'Fan:SystemModel')
```
Comment (Myoldmopar) [inline src/ExpandObjects/epfilter.f90:23689]: :+1: :+1: :+1:

File: src/ExpandObjects/epfilter.f90 [hunk 4]
```diff
@@ -31446,10 +31523,12 @@ SUBROUTINE CreateDOAS
       CALL AddToObjStr('Run on Sensible Load','Yes')
       ! CR8813 - MJW - Delete isHumidifierNone IF (.NOT. isDehumidifyNone .OR. .NOT. isHumidifierNone) THEN
       IF (.NOT. isDehumidifyNone) THEN
-        CALL AddToObjStr('Run on Latent Load','Yes',.TRUE.)
+        CALL AddToObjStr('Run on Latent Load','Yes')
```
Comment (Myoldmopar) [inline src/ExpandObjects/epfilter.f90:31526]: Not sure what the .TRUE. argument was, but OK.

Comment (Myoldmopar) [reply src/ExpandObjects/epfilter.f90:31526]: Oh, I see, is-last-item.  OK.

**Review by Myoldmopar — APPROVED** [review]
Code changes look great.  I'll test locally but it looks to be in great shape.

**Comment by Myoldmopar** [comment]
- Locally pulled develop into this branch and built, went fine.
- Ran all unit tests just for fun (I know this only touches ExpandObjects) and everything is clean
- Took `HVACTemplate-5ZoneConstantVolumeChillerBoiler.idf` file and tested with it:
  - Ran ExpandObjects on the base file: It went perfectly fine -- it replaced the building surface ground temperatures (@mjwitte is that expected?  I don't know a lot about this exact input file.)
  - Added an outdoor air schedule name in the HVACTemplate object and re-ran ExpandObjects.  I diffed the two expanded files and found that the newly expanded file has the schedule name reflected in the OA controller.
- Ran ExpandObjects on `HVACTemplate-5ZonePTAC-DOAS.idf` and found that the resulting coil system does now include the `Use Outdoor Air DX Coil` field
- Ran ExpandObjects on `HVACTemplate-5ZoneFanCoil-DOAS.idf` and found that it is using the system model.

CI looks overall good, but I noticed there was a unit test failure.  

```
 RUN ] ZoneUnitarySysTest.Test_UnitarySystemModel_factory
unknown file: Failure
C++ exception with description "Unable to determine fan air flow rate." thrown in the test body.
[ FAILED ] ZoneUnitarySysTest.Test_UnitarySystemModel_factory (191 ms)
```

I went ahead and deleted that build result and will let Mac re-run to verify it wasn't a fluke.  If it is still a problem, it will obviously have to be addressed before merging.  Maybe it'll be OK....but if not I guess maybe ExpandObjects isn't writing something it should be writing for the fan system model changes?

**Comment by Myoldmopar** [comment]
I had already run the unit tests but I went ahead and confirmed that test works perfectly for me:

```
$ ./energyplus_tests --gtest_filter=*Test_UnitarySystemModel_factory*
Note: Google Test filter = *Test_UnitarySystemModel_factory*
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from ZoneUnitarySysTest
[ RUN      ] ZoneUnitarySysTest.Test_UnitarySystemModel_factory
[       OK ] ZoneUnitarySysTest.Test_UnitarySystemModel_factory (50 ms)
[----------] 1 test from ZoneUnitarySysTest (50 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (131 ms total)
[  PASSED  ] 1 test.
```

**Comment by mjwitte** [comment]
Yes, the building surface ground temps get touched because expandobjects alos manages the GroundHeatTransfer:* objects which run the slab or basement preprocessors.  It probably shouldn't touch it if there aren't any GroundHeatTransfer:* objects, but it doesn't hurt anything.

**Comment by Myoldmopar** [comment]
@mjwitte so when I ran CI on that commit above, it showed a couple table-diff failures.  I was able to reproduce them locally and am adding another unit test to the regression scripts/utility now.  In the meantime, pulling develop into this branch allowed the table diffs to work properly, so there shouldn't be any danger in merging this in.  If you see any stray table diff issues in the meantime, pulling develop into that branch and letting CI re-run should be a good temporary workaround until I get the regression scripts fixed up.  I'm going to go ahead and merge this.

**Pull request merged**